gcd/src/main/java/org/spine3/server/storage/datastore/tenant/NamespaceAccess.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright 2017, TeamDev Ltd. All rights reserved.
+ *
+ * Redistribution and use in source and/or binary forms, with or without
+ * modification, must retain the above copyright notice and the following
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package org.spine3.server.storage.datastore.tenant;
+
+import com.google.cloud.datastore.Datastore;
+import com.google.cloud.datastore.Key;
+import com.google.cloud.datastore.Query;
+import com.google.common.base.Function;
+import com.google.common.collect.Iterators;
+import org.spine3.server.storage.datastore.Kind;
+import org.spine3.server.tenant.TenantIndex;
+import org.spine3.users.TenantId;
+
+import javax.annotation.Nullable;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Strings.isNullOrEmpty;
+
+/**
+ * A DAO for the Datastore {@link Namespace Namespaces}.
+ *
+ * @author Dmytro Dashenkov
+ */
+class NamespaceAccess implements TenantIndex {
+
+    private static final Kind NAMESPACE_KIND = Kind.ofNamespace();
+
+    private final Datastore datastore;
+
+    private final Set<Namespace> cache = new HashSet<>();
+
+    NamespaceAccess(Datastore datastore) {
+        this.datastore = datastore;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * <p>If the ID is not found, writes the ID into the in-mem cache.
+     *
+     * @param id the ID to ensure
+     */
+    @Override
+    public void keep(TenantId id) {
+        checkNotNull(id);
+        cache.add(Namespace.of(id));
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Set<TenantId> getAll() {
```

Comment:
There are potential concurrency issues with `fetchNamespaces()`, `exists()` and `getAll()`.